Temporal Nexus: Main Encyclopedia (#3053)
temporalio/documentation: docs/encyclopedia/temporal-nexus-security.mdx

---
id: nexus-security
title: Security in Nexus
sidebar_label: Security
description: Explore the security aspects of Temporal Nexus, including managing Nexus Endpoints, runtime access control, secure routing, and payload encryption to ensure safe and reliable cross-namespace operations.
toc_max_heading_level: 4
slug: /temporalnexus/security
keywords:
  - Nexus security
  - access control
  - encryption
tags:
  - security
  - temporal
  - distributed systems
---

### Security {#security}

#### Managing Nexus Endpoints in the Nexus API Registry

> Like `Namespaces`, a Nexus `Endpoint` is an account-scoped resource that is global within your Temporal Cloud account.

Reviewer: Again, we should explain this in the context of an open source deployment and Cloud.

Any `Developer` in an account, who is also a `Namespace Admin` on the endpoint’s target Namespace, can manage (create/update/delete) a Nexus `Endpoint`.\
All users with a `Read-only` role or higher in an account, can view and browse the full list of Endpoints.

#### Runtime Access Control {#runtime-access-control}

Access to a Nexus `Endpoint` is controlled by an allowlist policy (of caller Namespaces) for each endpoint in the Nexus API registry.

> Workers in each Namespace may authenticate with Temporal Cloud as they do today with mTLS certificates or the upcoming API key functionality. Once a worker has authenticated it can send Nexus `Operation` commands to Temporal Cloud using a Temporal SDK to start a Nexus `Operation` in a different Namespace. For example in the Temporal Go SDK a caller Workflow would use `nexusClient.ExecuteOperation` to issue a command to start a Nexus `Operation`.

Reviewer 2: ```suggestion
Workers in each Namespace may authenticate with Temporal Cloud as they do today with mTLS certificates or the upcoming API key functionality. 
Once a worker has authenticated it can send Nexus `Operation` commands to Temporal Cloud using a Temporal SDK to start a Nexus `Operation` in a different Namespace. 
For example, in the Temporal Go SDK a caller Workflow would use `nexusClient.ExecuteOperation` to issue a command to start a Nexus `Operation`.
```
One sentence per line

> Workers in each Namespace may authenticate with Temporal Cloud as they do today with mTLS certificates or the upcoming API key functionality. Once a worker has authenticated it can send Nexus `Operation` commands to Temporal Cloud using a Temporal SDK to start a Nexus `Operation` in a different Namespace. For example in the Temporal Go SDK a caller Workflow would use `nexusClient.ExecuteOperation` to issue a command to start a Nexus `Operation`.

Reviewer: Missing capitalization of proper nouns.

> To process a `ScheduleNexusOperation` command from a caller Workflow, Temporal Cloud obtains the handler Namespace and task-queue for the handler endpoint, and restricts access by verifying that the caller’s Namespace is in the endpoint’s allowlist.

Reviewer: Please stay consistent with casing, e.g: "Task Queue", "Endpoint".

> In this way, Temporal Cloud acts as a trusted broker across Namespace boundaries, and relies on authenticated workers in each Namespace. There is currently no provision for a handler’s endpoint to require a specific level of caller worker authentication, so either mTLS or API keys will be allowed as defined in the caller Namespace configuration. Nexus is designed to support additional Nexus access control policies in the future.

Reviewer 2: ```suggestion
In this way, Temporal Cloud acts as a trusted broker across Namespace boundaries, and relies on authenticated workers in each Namespace. 
There is currently no provision for a handler’s endpoint to require a specific level of caller worker authentication, so either mTLS or API keys will be allowed as defined in the caller Namespace configuration. 
Nexus is designed to support additional Nexus access control policies in the future.
```

Reviewer: I would probably just remove the "future" references and mTLS / API key mentions here and move the point about Cloud being a trusted broker to the paragraph above.

Author: Agreed.

#### Secure Routing {#secure-routing}

In Temporal Cloud multiple security provisions are in place to ensure it can act as a trusted broker across Namespace boundaries:

- Workers authenticate to their Namespaces via mTLS or an API key as allowed by their Namespace configuration.
- mTLS is used for all Nexus communication, including across cloud cells and regions, to:
  - Start or Cancel a `Nexus Operation`.
  - Callback on completion of an asynchronous `Nexus Operation`.
- Nexus `Endpoints` are only privately accessible from within a Temporal Cloud account
  - Accessible from within a caller Workflow using the Temporal SDK

>   - Not externally accessible for arbitrary clients yet.

Reviewer: You're inconsistent with the use of periods at the end of bullet points throughout the entire document. What does our style guide recommend here? Let's try to be consistent across the entire docs site.

- Future support for external Nexus `Endpoints` may have additional security measures such as:

>   - Signed callback URLs with single-use tokens to prevent forgery and attenuate permissions, for use over public networks.

Reviewer: It'll probably be encrypted, not signed and this applies not just for external endpoints.
I wouldn't mention future here either, it's confusing and we shouldn't commit to future stuff in the docs.

Author: Agreed.

#### Payload Encryption & Data Converter {#payload-encryption-&-data-converter}

> The `DataConverter` works the same for a Nexus `Operation` as it does for other payloads sent between a Worker and Temporal Cloud. Currently there is support for a single `DataConverter` to be used for inbound inputs and outbound outputs, which means the encryption key often used in a custom `DataConverter` must be the same for both the caller and the handler.

Reviewer 2: ```suggestion
The `DataConverter` works the same for a Nexus `Operation` as it does for other payloads sent between a Worker and Temporal Cloud. 
Currently there is support for a single `DataConverter` to be used for inbound inputs and outbound outputs, which means the encryption key often used in a custom `DataConverter` must be the same for both the caller and the handler.
```

Reviewer: I think Payloads is a proper noun, and should probably be a link.

> While there is currently no provision for using separate asymmetric encryption keys for each Namespace, we are open to adding support if required, so please reachout if you need this. This may require the exchange of public encryption keys across Namespaces, which could be inband (perhaps via the service registry) or out-of-band via existing org-specific infrastructure. Regardless, if separate asymmetric encryption keys are used the input for a Nexus `Operation` must be encrypted by the caller in a manner it can be decrypted by the handler and vice versa for the outputs, which would require changes to how the `DataConverter` works today.

Reviewer 3: ```suggestion
While there is currently no provision for using separate asymmetric encryption keys for each Namespace, we are open to adding support if required, so please contact us via XYZ XYZ if you need this. This may require the exchange of public encryption keys across Namespaces, which could be inband (perhaps via the service registry) or out-of-band via existing org-specific infrastructure. Regardless, if separate asymmetric encryption keys are used the input for a Nexus `Operation` must be encrypted by the caller in a manner it can be decrypted by the handler and vice versa for the outputs, which would require changes to how the `DataConverter` works today.
```

I would change this to an actual contact link now that it's going into Docs, eg like we have in https://docs.temporal.io/production-deployments/migration#multi-cluster-replication

Reviewer 2: One line per sentence

Reviewer: I wouldn't even mention reaching out to us in the public docs. And probably would just avoid mentioning unimplemented features as a general guideline.

Author: Agreed.

- [Managing Nexus Endpoints in the Nexus API Registry](#managing-nexus-endpoints)
- [Runtime Access Control](#runtime-access-control)
- [Secure Routing](#secure-routing)
- [Payload Encryption & Data Converter](#payload-encryption-data-converter)

### Managing Nexus Endpoints in the Nexus API Registry {#managing-nexus-endpoints}

Similar to Namespaces, a Nexus Endpoint is an account-scoped resource that is global within your Temporal Cloud account.

Any Developer in an account, who is also a Namespace Admin on the endpoint’s target Namespace, can manage (create/update/delete) a Nexus Endpoint.

All users with a Read-only role or higher in an account, can view and browse the full list of Endpoints.

### Runtime Access Control {#runtime-access-control}

Access to a Nexus Endpoint is controlled by an allowlist policy (of caller Namespaces) for each endpoint in the Nexus API registry.

Workers in each Namespace may authenticate with Temporal Cloud as they do today with mTLS certificates or the upcoming API key functionality. 
Once a Worker has authenticated it can send Nexus Operation commands to Temporal Cloud using a Temporal SDK to start a Nexus Operation in a different Namespace. 

> For example in the Temporal Go SDK a caller Workflow would use `nexusClient.ExecuteOperation` to issue a command to start a Nexus Operation.

Reviewer 2: ```suggestion
For example, in the Temporal Go SDK a caller Workflow would use `nexusClient.ExecuteOperation` to issue a command to start a Nexus Operation.
```
